File: non-empirical-figures/tradeoff_plane_schematic.py
#!/usr/bin/env python3
"""
Generate a schematic "throughput–FDR plane" figure in (q_H, L(eps)) coordinates.

Outputs:
  - tex/v8_figures/tradeoff_plane_schematic.pdf
  - tex/v8_figures/tradeoff_plane_schematic.png
"""

from __future__ import annotations

import os
from pathlib import Path
import tempfile


def main() -> None:
    # Put Matplotlib cache in a writable directory (avoids warnings on some systems).
    mpl_config_dir = Path(tempfile.gettempdir()) / "mplconfig-scientific-competition-overleaf"
    mpl_config_dir.mkdir(parents=True, exist_ok=True)
    os.environ["MPLCONFIGDIR"] = str(mpl_config_dir)
    os.environ.setdefault("XDG_CACHE_HOME", str(mpl_config_dir))

    import numpy as np
    import matplotlib

    matplotlib.use("Agg")
    import matplotlib.pyplot as plt
    from matplotlib.ticker import FuncFormatter

    out_dir = Path(__file__).resolve().parent
    pdf_path = out_dir / "tradeoff_plane_schematic.pdf"
    png_path = out_dir / "tradeoff_plane_schematic.png"

    plt.rcParams.update(
        {
            "font.size": 9,
            "font.family": "serif",
            "mathtext.fontset": "stix",
            "axes.linewidth": 0.8,
            "axes.labelpad": 2.5,
            "xtick.major.width": 0.8,
            "ytick.major.width": 0.8,
            "xtick.major.size": 3.0,
            "ytick.major.size": 3.0,
        }
    )

    # Schematic master bound: q_H * L <= K.
    # (Parameters chosen for visual clarity; no quantitative claim.)
    q_h = np.linspace(0.08, 1.0, 600)
    K = 4.0
    L_max = K / q_h

    fig, ax = plt.subplots(figsize=(6.6, 2.6), constrained_layout=True)
    ax.fill_between(q_h, 0.0, L_max, color="0.92", zorder=1)
    ax.plot(q_h, L_max, color="black", lw=2.6, zorder=3, label=r"Universal bound ($q_H\,L \lesssim \kappa$)")

    # Illustrative slices.
    L_target = 2.0
    ax.axhline(L_target, color="#2b6cb0", lw=1.8, ls="--", zorder=2, label=r"Fixed FDR target ($L$ fixed)")
    q_target = 0.35
    ax.axvline(q_target, color="0.55", lw=1.8, ls="--", zorder=2, label=r"Fixed recall ($q_H$ fixed)")

    # Illustrative RC attainment point on the boundary.
    q_rc = 0.5
    L_rc = K / q_rc
    ax.plot([q_rc], [L_rc], marker="o", ms=4.5, color="black", zorder=4)
    ax.text(
        q_rc + 0.04,
        L_rc - 0.35,
        "RC\n(attains bound)",
        fontsize=8.0,
        ha="left",
        va="top",
        bbox={"facecolor": "white", "edgecolor": "none", "alpha": 0.85, "pad": 1.5},
        zorder=5,
    )

    ax.legend(
        loc="upper right",
        frameon=False,
        fontsize=8.3,
        handlelength=2.8,
        labelspacing=0.35,
        borderaxespad=0.4,
    )

    ax.set_xlim(0.0, 1.0)
    ax.set_ylim(0.0, 8.5)
    ax.set_xlabel(r"Recall $q_H$")
    ax.set_ylabel(r"Purity stringency $L(\varepsilon)=\log(1/\eta(\varepsilon))$")
    ax.set_xticks([0.0, 0.5, 1.0])
    ax.set_yticks([0.0, 2.0, 4.0, 6.0, 8.0])
    fmt = FuncFormatter(lambda x, _pos: f"{x:g}")
    ax.xaxis.set_major_formatter(fmt)
    ax.yaxis.set_major_formatter(fmt)
    ax.spines["top"].set_visible(False)
    ax.spines["right"].set_visible(False)

    fig.savefig(pdf_path, bbox_inches="tight", pad_inches=0.02)
    fig.savefig(png_path, dpi=300, bbox_inches="tight", pad_inches=0.02)
    print(f"Wrote {pdf_path}")
    print(f"Wrote {png_path}")


if __name__ == "__main__":
    main()

File: non-empirical-figures/tradeoff_plane_schematic.jl
using Plots, LaTeXStrings
import PyPlot

pyplot()
PyPlot.rc("text", usetex=true)
PyPlot.rc("font", family="serif")
PyPlot.matplotlib.rcParams["mathtext.fontset"] = "cm"
PyPlot.matplotlib.rcParams["text.latex.preamble"] = "\\usepackage{amssymb}"

# Generate data for master bound: q_H * L <= K
# Start q_h where the bound enters the visible region (L = 13 → q_h = K/13 ≈ 0.31)
q_h = range(0.30, 1.0, length=600)
K = 4.0
L_max = K ./ q_h

# Colors
bound_color = :black
fdr_color = colorant"#2b6cb0"
recall_color = colorant"#8c8c8c"

# Illustrative slice parameters
L_target = 2.0
q_target = 0.35

# RC attainment point on the boundary
q_rc = 0.5
L_rc = K / q_rc

# Create figure with fill under curve
fig = plot(collect(q_h), L_max,
    fillrange = 0.0,
    fillcolor = colorant"#ededed",
    linecolor = nothing,
    label = false,
    size = (1980, 780),
    dpi = 300,
    xlabel = L"Recall $q_H$",
    ylabel = L"Purity stringency $L(\varepsilon)=\log(1/\eta(\varepsilon))$",
    xlims = (0.0, 1.0),
    ylims = (0.0, 13.0),
    xticks = [0.0, 0.5, 1.0],
    yticks = [0.0, 4.0, 8.0, 12.0],
    legend = false,
    tickfontsize = 12,
    guidefontsize = 24,
    grid = false,
)

# Plot master bound curve
plot!(fig, collect(q_h), L_max,
    color = bound_color,
    linewidth = 3,
    label = L"Universal bound ($q_H\,L \leq \kappa$)")

# Plot fixed FDR target (horizontal dashed line)
hline!(fig, [L_target],
    color = fdr_color,
    linewidth = 2.5,
    linestyle = :dash,
    label = L"Fixed FDR target ($L$ fixed)")

# Plot fixed recall (vertical dashed line)
vline!(fig, [q_target],
    color = recall_color,
    linewidth = 2.5,
    linestyle = :dash,
    label = L"Fixed recall ($q_H$ fixed)")

# Plot RC attainment point
scatter!(fig, [q_rc], [L_rc],
    color = :black,
    markersize = 12,
    markerstrokewidth = 0,
    label = false)

# Add text annotation for RC point
annotate!(fig, q_rc + 0.06, L_rc - 0.5, text("RC\n(attains bound)", 18, :black, :left))

# Render the Plots.jl figure to PyPlot
display(fig)

# Now access the PyPlot figure and axes
pyfig = PyPlot.gcf()
ax = PyPlot.gca()

# Remove top and right spines
ax.spines["top"].set_visible(false)
ax.spines["right"].set_visible(false)

# Create legend handles manually
using PyPlot: matplotlib
line_bound = matplotlib.lines.Line2D([0], [0], color="black", linewidth=3, label=L"Universal bound ($q_H\,L \leq \kappa$)")
line_fdr = matplotlib.lines.Line2D([0], [0], color="#2b6cb0", linewidth=2.5, linestyle="--", label=L"Fixed FDR target ($L$ fixed)")
line_recall = matplotlib.lines.Line2D([0], [0], color="#8c8c8c", linewidth=2.5, linestyle="--", label=L"Fixed recall ($q_H$ fixed)")

handles = [line_bound, line_fdr, line_recall]

# Position legend at upper right
ax.legend(handles=handles, loc="upper right", bbox_to_anchor=(1.0, 1.0),
          fontsize=24, frameon=false)

# Save figures
PyPlot.savefig("non-empirical-figures/tradeoff_plane_schematic.pdf", bbox_inches="tight", facecolor="white")

# Save in overleaf folder
#pyfig.savefig("../overleaf/tex/v8_figures/tradeoff_plane_schematic.pdf", bbox_inches="tight", transparent=true)
